src/bsfs.c

```diff
@@ -551,6 +553,87 @@ int bsfs_ftruncate(bs_file_t file, off_t size) {
   return -ENOSYS;
 }
 
+static bool get_timestamp(const struct timespec times[2], size_t index,
+                          bft_timestamp_t* out) {
+  if (!times) {
+    *out = time(NULL);
+    return true;
+  }
+
+  const struct timespec* ret_time = times + index;
+  if (ret_time->tv_nsec == UTIME_OMIT) {
+    return false;
+  }
+
+  if (ret_time->tv_nsec == UTIME_NOW) {
+    *out = time(NULL);
+    return true;
+  }
+
+  *out = ret_time->tv_sec;
+  return true;
+}
+
+static int do_utimens(bs_open_level_t level, bft_offset_t index,
+                      const struct timespec times[2]) {
+  bft_timestamp_t atim; // Access time
+  bft_timestamp_t mtim; // Modification time
+  bool needs_atim = get_timestamp(times, 0, &atim);
+  bool needs_mtim = get_timestamp(times, 1, &mtim);
+
+  if (!needs_atim && !needs_mtim) {
+    return 0;
+  }
+
+  bft_entry_t ent;
+  int ret = bft_read_table_entry(level->bft, &ent, index);
```

Comment:
I would add a comment that makes clear that the metadata is locked.

Reply:
The convention so far has been that any function starting with `do_` assumes that metadata is locked. We should probably document that somewhere.

Reply:
I wasn't sure about that. If that is the case you can merge the branch with master.